### ISSUE TITLE
False Positive `E1130: bad operand type for unary ~: object (invalid-unary-operand-type)`

### ISSUE THREAD

**Issue by astjohn-pilot** [PyCQA/pylint#8505]
### Bug description

# False Positive `E1130: bad operand type for unary ~: object (invalid-unary-operand-type)`

# Sample Code to Reproduce False Positive
```python
"""simple testing file for issue report to pylint"""
from functools import reduce
from pyspark.sql.functions import col, when
from pyspark.sql import SparkSession

spark = SparkSession.builder.getOrCreate()

df = spark.createDataFrame(
    [(True, False, False)],
    ["bool_col1", "bool_col2", "bool_col3"],
)


bool_cols = [
    "bool_col1",
    "bool_col2",
    "bool_col3",
]
conditions = reduce(lambda a, b: a | b, [col(c) for c in bool_cols])

df.withColumn("one_col_is_true", when(conditions, True).otherwise(False)).withColumn(
    "no_col_is_true", when(~conditions, True).otherwise(False)
).show()
```
# Output
```text
+---------+---------+---------+---------------+--------------+
|bool_col1|bool_col2|bool_col3|one_col_is_true|no_col_is_true|
+---------+---------+---------+---------------+--------------+
|     true|    false|    false|           true|         false|
+---------+---------+---------+---------------+--------------+
```

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:22:27: E1130: bad operand type for unary ~: object (invalid-unary-operand-type)

------------------------------------------------------------------
Your code has been rated at 3.75/10 (previous run: 3.75/10, +0.00)
```

### Expected behavior

No `invalid-unary-operand-type` error.

### Pylint version

```shell
pylint 2.17.1
astroid 2.15.1
Python 3.10.9 | packaged by conda-forge | (main, Jan 11 2023, 15:15:40) [MSC v.1916 64 bit (AMD64)]
```

### OS / Environment

Windows 10

### Additional dependencies

pyspark==3.3.0
